File: 84.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        maxArea = -1
        n = len(heights)
        left = [0] * n
        right = [0] * n
        stack = []
        for i in range(len(heights)):
            if not stack:
                stack.append(i)
                left[i] = 0
            else:
                while stack and heights[stack[-1]] >= heights[i]:
                    stack.pop()
                if not stack:
                    left[i] = 0
                else:
                    left[i] = stack[-1] + 1
                stack.append(i)
        stack.clear()
        for i in range(len(heights) - 1, -1, -1):
            if not stack:
                stack.append(i)
                right[i] = n - 1
            else:
                while stack and heights[stack[-1]] >= heights[i]:
                    stack.pop()
                if not stack:
                    right[i] = n - 1
                else:
                    right[i] = stack[-1] - 1
                stack.append(i)
        for i in range(n):
            maxArea = max(maxArea, (right[i] - left[i] + 1) * heights[i])
        return maxArea


print(Solution().largestRectangleArea([2, 1, 5, 6, 2, 3]))
print("expected 10")

# brute force
# def largestRectangleArea(self, heights: list[int]) -> int:
#     maxArea = -1
#     for idx, height in enumerate(heights):
#         l = r = idx
#         while l - 1 > -1 and heights[l - 1] >= height:
#             l -= 1
#         while r + 1 < len(heights) and heights[r + 1] >= height:
#             r += 1
#         maxArea = max(maxArea, (r - l + 1) * height)
#     return maxArea

File: 84.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
private:
  struct Datum {
    size_t startIndex;
    int height;
  };

public:
  int largestRectangleArea(vector<int> &heights) {
    vector<Datum> stack;
    int maxArea = 0;
    for (size_t i = 0; i < heights.size(); ++i) {
      size_t start = i;
      while (!stack.empty() && stack.back().height > heights[i]) {
        size_t startIndex = stack.back().startIndex;
        int height = stack.back().height;
        stack.pop_back();
        maxArea = max(maxArea, (int)(i - startIndex) * height);
        start = startIndex;
      }
      stack.push_back({start, heights[i]});
    }
    for (int i = 0; i < stack.size(); ++i) {
      size_t index = stack[i].startIndex;
      int height = stack[i].height;
      maxArea = max(maxArea, (int)(heights.size() - index) * height);
    }
    return maxArea;
  }
};
int main() {
  vector<int> t1_in = {2, 1, 5, 6, 2, 3};
  int t1_res = Solution().largestRectangleArea(t1_in);
  cout << t1_res << endl;
  cout << "expected 10" << endl;
}